### PR TITLE
Minimal spi master

### LINKED ISSUE #213: Hardware support to access flash
In some future product, not necessarily all versions of the TKey, we would like to be able to access the onboard flash chip from the CPU, perhaps only by the firmware, possibly also from device apps.

The first need is to support a pre-loaded app on the flash on some products.

This issue includes these

- [ ] #234 
- [x] #207

### PR DESCRIPTION
**Pull request by dehanj**
This supersedes #203

> This PR adds a minimal SPI master to the application_fpga. This allows applications to access the Flash memory on the Tkey. The SPI master basically supports generating a clock, control the chip select, and send and receive a byte. SW much implement the functionality required to send commands and receive responses. Updates to tkey_libs will add this functionality.
> 
> Currently the SPI master is enabled in FPGA builds by including the flag symbol INCLUDE_SPI_MASTER to Yosys in the Makefile. The Makefile does not yet change pin constraint file. Instead the constraint file now always includes pins for the SPI interface.

Rebased on main, and squashed.
- SPI master construction (also added a job to CI)
- Testbench and simulation

Closes #213

### CODE REVIEW

**Comment by dehanj** [comment]
To minimize the risk during rebase and squash I created a new branch and a new PR.
I suggest we merge this isntead of https://github.com/tillitis/tillitis-key1/pull/203 when we are ready and have tested it properly.

**Comment by dehanj** [comment]
I have verified both the builds with and without the SPI master. So far it looks good.

**Comment by dehanj** [comment]
Added SPI build to CI. Also divided the builds to different jobs, to make it easier to see what fails or not.

**Comment by dehanj** [comment]
This PR should be merged after #208, since it is a risk here that cache fetches the wrong bitstream.

File: hw/application_fpga/Makefile [hunk 1]
```diff
@@ -236,9 +237,18 @@ tb:
 #-------------------------------------------------------------------
 # Main FPGA build flow.
 # Synthesis. Place & Route. Bitstream generation.
+#
```
Comment (secworks) [inline hw/application_fpga/Makefile:240]: Är detta för ny target. Annar bode den texten flyttas till beskrivning av ny target och formuleras om.

File: hw/application_fpga/Makefile [hunk 2]
```diff
@@ -236,9 +237,18 @@ tb:
...
+# To include the SPI-master, add the flag -DINCLUDE_SPI_MASTER to Yosys cmd.
+# This can, for example, be done using
+# 'make application_fpga.bin YOSYS_FLAG=-DINCLUDE_SPI_MASTER'.
+# Remember to do a make clean between builds with and wihtout the SPI master,
+# otherwise there is a risk of including the it unknowingly.
```
Comment (secworks) [inline hw/application_fpga/Makefile:245]: 'the it'. Remove 'it'.

**Review by secworks — CHANGES_REQUESTED** [review]

File: hw/application_fpga/Makefile (hunk 2 above)
Comment (secworks) [reply hw/application_fpga/Makefile:245]: Or change 'it' to 'SPI master'.

Comment (dehanj) [reply hw/application_fpga/Makefile]: Fixed.

File: hw/application_fpga/core/tk1/tb/tb_tk1.v [hunk 3]
```diff
@@ -58,6 +58,10 @@ module tb_tk1();
   localparam ADDR_CPU_MON_FIRST = 8'h61;
   localparam ADDR_CPU_MON_LAST  = 8'h62;
 
+  localparam ADDR_SPI_EN        = 8'h80;
```
[the remark this answers is not in the document]
Comment (dehanj) [reply hw/application_fpga/core/tk1/tb/tb_tk1.v:61]: I don't have the answer if we should use ifdef in the tb as well. One argument is that it is easier to have on build and make sure it all works. But on the other hand, if it is substantial difference and it is a risk that the test fails, yeah then it is better to have two.

But I think you will have to answer for what is the best approach here Joachim.

File: hw/application_fpga/core/tk1/tb/tb_tk1.v [hunk 4]
```diff
@@ -89,13 +93,24 @@ module tb_tk1();
   wire          tb_gpio3;
   wire          tb_gpio4;
 
+  wire          tb_spi_ss;
```
[the remark this answers is not in the document]
Comment (dehanj) [reply hw/application_fpga/core/tk1/tb/tb_tk1.v:96]: See my response in the other comment.

[the remark this answers is not in the document]
Comment (secworks) [reply hw/application_fpga/core/tk1/tb/tb_tk1.v:96]: Ok.

File: hw/application_fpga/core/tk1/tb/tb_tk1.v (hunk 3 above)
[the remark this answers is not in the document]
Comment (secworks) [reply hw/application_fpga/core/tk1/tb/tb_tk1.v:61]: Having excess symbolic names should pose not problem. A linter may give warnings (This is not Go after all).
But ports in the DUT and associated wires should cause at least warnings during build of the som environment, and cause confusion. Testbench functionality such a state monitor, state display may also become problematic.

And most importantly. The test cases in a testbench should test the functionality of the DUT. This means that the DUT and the TB are in sync in terms of functionality. A self-checking test case that tests a functionality that don't exist in the DUT should fail.

To summarize. If we have functionality in the DUT (the core) that is compile time optional, we should also have the test functionality for that functionality compile time optional.

[the remark this answers is not in the document]
Comment (dehanj) [reply hw/application_fpga/core/tk1/tb/tb_tk1.v:61]: Sounds great. Can you add it?

**Comment by mchack-work** [comment]
SPI control addresses were missing from tk1_mem.h. I added them.

**Review by mchack-work — APPROVED** [review]
